File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        """
        [1, 2, 3, 4]
               i
               
        prefixes:
        [1, 1, 2, 6]
        suffixes:
        [24, 12, 4, 1]
        
        create a prefixes array that holds the product of all indices before the curr index
        do the same for suffixes/after the curr index
        now, accessing the prefix/suffix array at i will give you the product of all prev/future ele's
        loop through nums
        multiply the prefix and suffix for that index together 
        push that product to a return array
        return array after iterating through nums
        """
        
        prefixes = [1]
        i = 1
        while i < len(nums):
            product = prefixes[i - 1] * nums[i - 1]
            prefixes.append(product)
            i += 1
        print('pres', prefixes)
        
        suffixes = [0] * len(nums)
        suffixes[-1] = 1
        i = len(nums) - 2
        while i >= 0:
            suffixes[i] = suffixes[i + 1] * nums[i + 1]
            i -= 1
        print('sufs', suffixes)   
            
        rtn = []
        for i in range(len(nums)):
            rtn.append(prefixes[i] * suffixes[i])
            
        return rtn
        
        

File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        // return bruteForce(nums);
        return something(nums);
    }
    
    public int[] something(int[] nums) {
        /*
            the product of all but curr num is the same as the
            product of all to the left and all to the right of curr.
            if you make two passes, you can fill up two different arrays
            that will store the "left prod" and "right prod" for every idx.
            
            L: [1*,   2, 8, 48] 
            R: [192, 48, 8, 1*]
                      i
               [2,    4, 6,  8]
        */
        
        int[] leftProds = new int[nums.length];
        int[] rightProds = new int[nums.length];
        leftProds[0] = 1;
        rightProds[rightProds.length - 1] = 1;
        
        for (int i = 1; i < nums.length; ++i)
            leftProds[i] = leftProds[i - 1] * nums[i - 1];
        
        for (int i = nums.length - 2; i >= 0; --i)
            rightProds[i] = rightProds[i + 1] * nums[i + 1];
        
        int[] answer = new int[nums.length];
        for (int i = 0; i < nums.length; ++i)
            answer[i] = leftProds[i] * rightProds[i];
    
        return answer;
    }
    
    public int[] bruteForce(int[] nums) {
        int[] answer = new int[nums.length];
        
        for (int i = 0; i < nums.length; ++i) {
            int product = 1;
            for (int j = 0; j < nums.length; ++j) {
                if (j == i) continue;
                
                product *= nums[j];
            }
            
            answer[i] = product;
        }
        
        return answer;
    }
}